File: windows/src/Patchthrough.Core/ProjectVocabulary.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace Patchthrough.Core;

public static partial class ProjectVocabulary
{
    private const int MaximumTerms = 256;
    private static readonly string[] Manifests =
        ["package.json", "Package.swift", "Cargo.toml", "pyproject.toml", "requirements.txt", "go.mod", "Gemfile", "Podfile"];
    private static readonly HashSet<string> Ignored =
        new([".git", ".build", "build", "dist", "node_modules", "vendor", "Pods"], StringComparer.OrdinalIgnoreCase);

    public static IReadOnlyList<VocabularyTerm> Collect(string? projectRoot)
    {
        var terms = new Dictionary<string, VocabularyTerm>(StringComparer.OrdinalIgnoreCase);
        void Add(string text, string source, double weight)
        {
            var cleaned = text.Trim();
            if (terms.Count >= MaximumTerms || cleaned.Length is < 3 or > 64 || !cleaned.Any(char.IsLetter)) return;
            if (!terms.TryGetValue(cleaned, out var existing) || existing.Weight < weight)
                terms[cleaned] = new VocabularyTerm(cleaned, source, Math.Clamp(weight, 0, 2));
        }

        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        AddLines(Path.Combine(home, ".config", "patchthrough", "vocabulary.txt"), "personal_glossary", 2);
        if (projectRoot is null || !Directory.Exists(projectRoot)) return Sorted();
        AddLines(Path.Combine(projectRoot, ".patchthrough", "vocabulary.txt"), "project_glossary", 2);

        var glossary = Path.Combine(projectRoot, ".patchthrough", "vocabulary.json");
        try
        {
            var node = JsonNode.Parse(File.ReadAllText(glossary));
            var array = node as JsonArray ?? node?["terms"] as JsonArray;
            foreach (var value in array?.Select(item => item?.GetValue<string>()) ?? [])
                if (value is not null) Add(value, "project_glossary", 2);
        }
        catch (Exception error) when (error is IOException or JsonException) { }

        foreach (var manifest in Manifests)
        {
            var path = Path.Combine(projectRoot, manifest);
            if (!File.Exists(path) || new FileInfo(path).Length > 512_000) continue;
            foreach (Match match in IdentifierRegex().Matches(File.ReadAllText(path)))
                if (IsSpecialized(match.Value)) Add(match.Value, "manifest", 1.4);
        }

        var head = Path.Combine(projectRoot, ".git", "HEAD");
        if (File.Exists(head))
        {
            foreach (var part in File.ReadAllText(head).Split(new[] { '/', '-', '_' }, StringSplitOptions.RemoveEmptyEntries))
                Add(part, "branch", 0.8);
        }

        var count = 0;
        foreach (var path in Enumerate(projectRoot, 0))
        {
            if (++count > 400) break;
            var stem = Path.GetFileNameWithoutExtension(path);
            if (IsSpecialized(stem)) Add(stem, "filename", 0.8);
            if (!new[] { ".swift", ".cs", ".ts", ".tsx", ".js", ".jsx", ".rs", ".go", ".py" }
                .Contains(Path.GetExtension(path), StringComparer.OrdinalIgnoreCase)) continue;
            if (new FileInfo(path).Length > 128_000) continue;
            foreach (Match match in DeclarationRegex().Matches(File.ReadAllText(path)))
                Add(match.Groups[1].Value, "declaration", 1);
        }
        return Sorted();

        void AddLines(string path, string source, double weight)
        {
            if (!File.Exists(path)) return;
            foreach (var line in File.ReadLines(path).Where(line => !line.TrimStart().StartsWith('#')))
                Add(line, source, weight);
        }

        IReadOnlyList<VocabularyTerm> Sorted() => terms.Values
            .OrderByDescending(term => term.Weight).ThenBy(term => term.Text, StringComparer.OrdinalIgnoreCase)
            .Take(MaximumTerms).ToList();
    }

    private static IEnumerable<string> Enumerate(string directory, int depth)
    {
        if (depth > 3) yield break;
        IEnumerable<string> children;
        try { children = Directory.EnumerateFileSystemEntries(directory); }
        catch (UnauthorizedAccessException) { yield break; }
        foreach (var child in children)
        {
            if (Directory.Exists(child))
            {
                if (Ignored.Contains(Path.GetFileName(child))) continue;
                foreach (var nested in Enumerate(child, depth + 1)) yield return nested;
            }
            else yield return child;
        }
    }

    private static bool IsSpecialized(string term) =>
        term.Any(char.IsUpper) || term.Contains('-') || term.Contains('_') || term.Contains('.');

    [GeneratedRegex("[A-Za-z][A-Za-z0-9_.+\\-]{2,63}")]
    private static partial Regex IdentifierRegex();

    [GeneratedRegex("(?:class|struct|enum|protocol|interface|record|func|function)\\s+([A-Za-z][A-Za-z0-9_]{2,63})")]
    private static partial Regex DeclarationRegex();
}

File: Sources/patchthrough/Transcription/ProjectVocabulary.swift
import Foundation

enum ProjectVocabulary {
    private static let maximumTerms = 256
    private static let manifests = [
        "package.json", "Package.swift", "Cargo.toml", "pyproject.toml",
        "requirements.txt", "go.mod", "Gemfile", "Podfile",
    ]
    private static let ignored = Set([".git", ".build", "build", "dist", "node_modules", "vendor", "Pods"])
    private static let common = Set([
        "class", "const", "dependencies", "dependency", "description", "development",
        "function", "import", "interface", "package", "private", "project", "public",
        "return", "string", "struct", "target", "version",
    ])

    static func collect(projectRoot: URL?) -> [VocabularyTerm] {
        var terms: [String: VocabularyTerm] = [:]
        func add(_ text: String, source: String, weight: Double) {
            let cleaned = text.trimmingCharacters(in: .whitespacesAndNewlines)
            guard terms.count < maximumTerms,
                  cleaned.count >= 3, cleaned.count <= 64,
                  cleaned.rangeOfCharacter(from: .letters) != nil,
                  !common.contains(cleaned.lowercased()) else { return }
            let key = cleaned.lowercased()
            if terms[key] == nil || terms[key]!.weight < weight {
                terms[key] = VocabularyTerm(text: cleaned, source: source, weight: weight)
            }
        }

        let personal = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".config/patchthrough/vocabulary.txt")
        for line in lines(at: personal) { add(line, source: "personal_glossary", weight: 2) }

        guard let root = projectRoot, FileManager.default.fileExists(atPath: root.path) else {
            return Array(terms.values).sorted { $0.text.localizedCaseInsensitiveCompare($1.text) == .orderedAscending }
        }
        for line in lines(at: root.appendingPathComponent(".patchthrough/vocabulary.txt")) {
            add(line, source: "project_glossary", weight: 2)
        }
        let jsonGlossary = root.appendingPathComponent(".patchthrough/vocabulary.json")
        if let data = try? Data(contentsOf: jsonGlossary),
           let object = try? JSONSerialization.jsonObject(with: data) {
            let values = object as? [String] ?? (object as? [String: Any])?["terms"] as? [String] ?? []
            for value in values { add(value, source: "project_glossary", weight: 2) }
        }

        for manifest in manifests {
            let url = root.appendingPathComponent(manifest)
            guard let data = try? Data(contentsOf: url), data.count <= 512_000,
                  let text = String(data: data, encoding: .utf8) else { continue }
            for candidate in identifiers(in: text) where isSpecialized(candidate) {
                add(candidate, source: "manifest", weight: 1.4)
            }
        }

        let head = root.appendingPathComponent(".git/HEAD")
        if let value = try? String(contentsOf: head, encoding: .utf8),
           let branch = value.split(separator: "/").last {
            for part in branch.split(whereSeparator: { "-_".contains($0) }) {
                add(String(part), source: "branch", weight: 0.8)
            }
        }

        if let enumerator = FileManager.default.enumerator(
            at: root,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: [.skipsHiddenFiles, .skipsPackageDescendants]
        ) {
            var files = 0
            for case let file as URL in enumerator {
                if ignored.contains(file.lastPathComponent) { enumerator.skipDescendants(); continue }
                let depth = file.pathComponents.count - root.pathComponents.count
                if depth > 3 { enumerator.skipDescendants(); continue }
                guard (try? file.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) != true else { continue }
                files += 1
                if files > 400 { break }
                let stem = file.deletingPathExtension().lastPathComponent
                if isSpecialized(stem) { add(stem, source: "filename", weight: 0.8) }
                guard ["swift", "cs", "ts", "tsx", "js", "jsx", "rs", "go", "py"].contains(file.pathExtension),
                      let data = try? Data(contentsOf: file), data.count <= 128_000,
                      let source = String(data: data, encoding: .utf8) else { continue }
                for declaration in declarations(in: source) {
                    add(declaration, source: "declaration", weight: 1)
                }
            }
        }
        return Array(terms.values)
            .sorted { $0.weight != $1.weight ? $0.weight > $1.weight : $0.text < $1.text }
            .prefix(maximumTerms).map { $0 }
    }

    private static func lines(at url: URL) -> [String] {
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return [] }
        return text.split(whereSeparator: \.isNewline).map(String.init)
            .filter { !$0.trimmingCharacters(in: .whitespaces).hasPrefix("#") }
    }

    private static func identifiers(in text: String) -> [String] {
        matches(#"[A-Za-z][A-Za-z0-9_.+\-]{2,63}"#, in: text)
    }

    private static func declarations(in text: String) -> [String] {
        matches(#"(?:class|struct|enum|protocol|interface|record|func|function)\s+([A-Za-z][A-Za-z0-9_]{2,63})"#, in: text, group: 1)
    }

    private static func matches(_ pattern: String, in text: String, group: Int = 0) -> [String] {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }
        let range = NSRange(text.startIndex..., in: text)
        return regex.matches(in: text, range: range).compactMap { match in
            Range(match.range(at: group), in: text).map { String(text[$0]) }
        }
    }

    private static func isSpecialized(_ term: String) -> Bool {
        term.contains(where: \.isUppercase)
            || term.contains("-") || term.contains("_") || term.contains(".")
    }
}
